core/arch/arm/kernel/kern.ld.S

```diff
@@ -394,6 +394,7 @@ SECTIONS
 	__init_size = __data_end - CFG_TEE_LOAD_ADDR;
 	__init_mem_usage = __end - CFG_TEE_LOAD_ADDR;
 #endif
+	ASSERT(. <= (CFG_TEE_RAM_START + CFG_TEE_RAM_VA_SIZE), "the CFG_TEE_RAM_VA_SIZE is not enough!")
```

Comment:
A quick update on this... After reading the GNU ld documentation and experimenting a bit, I understand better why the following line (`. = CFG_TEE_RAM_START + CFG_TEE_RAM_VA_SIZE`) cannot generate an error even when `CFG_TEE_RAM_VA_SIZE` is too small, and therefore why the `ASSERT()` is useful.

The rule is not "the location counter can't ever go backwards". From the GNU ld documentation:
> The location counter may not be moved backwards inside an output section, and may not be moved backwards outside of an output section if so doing creates areas with overlapping LMAs.

Here we're not inside an output section. So `.` *can* be moved backwards as long as it does not cause a conflict. Since we're at the end of the sections list, the linker won't output any other section and therefore there can be no "overlapping LMA" due to the counter being re-positioned.